src/main/java/fr/sncf/osrd/infra/railscript/RSExpr.java

```diff
@@ -702,5 +701,45 @@ public void accept(RSExprVisitor visitor) throws InvalidInfraException {
         }
     }
 
+    
+    public static final class PreviousReservedRoute extends RSExpr<RSOptional<RouteState>> {
+        public final RSExpr<SignalState> signal;
+
+        public PreviousReservedRoute(RSExpr<SignalState> signal) {
+            this.signal = signal;
+        }
+
+        /**
+         * Evaluate the expression to an optional route
+         * @param state the global state of the program
+         * @return the current reserved route preceding the signal, if any. Else an empty optional
+         */
+        @Override
+        public RSOptional<RouteState> evaluate(RSExprState<?> state) {
+            var currentSignal = signal.evaluate(state).signal;
+            assert currentSignal.direction.directionSet.length == 1;
+            var routes = currentSignal.linkedDetector.getIncomingRouteNeighbors(
+                    currentSignal.direction.directionSet[0]
```

Comment:
signal.direction should have type EdgeDirection, not ApplicableDirection, as signals never are applicable on both directions (this isn't your mistake). you can either fix it by changing the type and raising an error during parsing, or just leaving it this way